pennylane/devices/qubit/simulate.py

```diff
@@ -44,6 +45,30 @@ def measure_state_diagonalizing_gates(
     return measurementprocess.process_state(qml.math.flatten(state), wires)
 
 
+def measure_hamiltonian_expval(
+    measurementprocess: MeasurementProcess, state: TensorLike
+) -> TensorLike:
+    """Measure the expectation value of the state when the measured observable is a ``Hamiltonian`` or
+    ``SparseHamiltonian``.
+
+    Args:
+        measurementprocess (MeasurementProcess): measurement process to apply to the state
+        state (TensorLike): the state to measure
+
+    Returns:
+        TensorLike: the result of the measurement
+    """
+    # Add case for backprop later
+    Hmat = measurementprocess.obs.sparse_matrix()
+    _state = qml.math.toarray(state)
+    res = csr_matrix.dot(
+        csr_matrix(qml.math.conj(_state)),
+        csr_matrix.dot(Hmat, csr_matrix(_state[..., None])),
+    ).toarray()[0]
```

Comment:
I know this is what qubit device does, but I'd prefer to split this out into multiple "lines" of code with explanatory variable names, not one line spread out over four. Each line should do one thing. It'll be much easier to read.